src/lib/support/StateMachine.h

```diff
@@ -0,0 +1,251 @@
+ *
+ * For event dispatch, the State Machine implements the Context interface.  The Context
+ * interface is passed to States to allow Dispatch() of events when needed.
+ *
+ * The State held in the TState must provide four methods to support calls from
+ * the State Machine:
+ * @code
+ *   struct State {
+ *     void Enter() { }
+ *     void Exit() { }
+ *     void LogTransition(const char *) { }
+ *     const char *GetName() { return ""; }
+ *   }
+ * @endcode
+ * @note The State::Enter() method is allowed to Dispatch events.  However, dispatch calls
+ * from State::Exit() are ignored.
+ *
+ * The TTransitions table type is implemented with an overloaded callable operator method
+ * to match the combinations of State / Event variants that may produce a new-state return.
+ * This allows the Transition table to define how each State responds to Events.  Below is
+ * an example of a Transitions table implemented as a struct:
+ *
+ * @code
+ *   struct Transitions {
+ *     using State = chip::StateMachine::VariantState<State1, State2>;
+ *     chip::StateMachine::Optional<State> operator()(State &state, Event &event)
+ *     {
+ *         if (state.Is<State1>() && event.Is<Event2>())
+ *         {
+ *             return State::Create<State2>();
+ *         }
+ *         else if (state.Is<State2>() && event.Is<Event1>())
+ *         {
+ *             return State::Create<State1>();
+ *         }
+ *         else
+ *         {
+ *             return {}
+ *         }
+ *     }
+ *   }
+ * @endcode
+ * @note The Transition table is allowed to Dispatch events.  However,
+ * dispatched events will be ignored if a new state transition is returned.
+ *
+ * @tparam TState a variant holding the States.
+ * @tparam TEvent a variant holding the Events.
+ * @tparam TTransitions an object that implements the () operator for transitions.
+ */
+template <typename TState, typename TEvent, typename TTransitions>
+class StateMachine : public Context<TEvent>
+{
+public:
+    StateMachine(TTransitions & tr) : mCurrentState(tr.GetInitState()), mTransitions(tr) {}
+    ~StateMachine() override = default;
+    void Dispatch(const TEvent & evt) override
+    {
+        auto inProcess = !events.empty();
+        events.push_back(evt);
+        if (!inProcess)
+        {
+            HandleEvents();
+        }
+    }
+
+    TState mCurrentState;
+
+private:
+    void HandleEvents()
+    {
+        while (!events.empty())
+        {
+            auto count    = events.size();
+            auto optState = mTransitions(mCurrentState, events.front());
+            if (optState.HasValue())
+            {
+                auto newState = optState.Value();
+                newState.LogTransition(mCurrentState.GetName());
+                mCurrentState.Exit();
+                mCurrentState = newState;
+                while (events.size() > count)
+                {
+                    events.pop_back(); // events discarded per design
```

Comment:
So does this code assume that `mTransitions` call can't add new events? If so, why exactly do we need the deque? Couldn't we just perform one iteration based on the event passed as an argument to `Dispatch`? Is this class going to be used on any, including constrained, devices? I'm not sure how much attention I should pay to memory usage while reviewing :)

Reply:
> So does this code assume that mTransitions call can't add new events?

No, the rules are slightly more complicated.  The mTransitions call can return a new state transition, OR it can emit 0, 1 or more events.  And State::Enter() can emit 0, 1 or more events, but State::Exit() is not permitted to emit events because it is nonsensical for a State to continue to drive the state machine once it is already exiting.

I think the more intuitive way to look at it is this: code is permitted to emit events from State::Enter() until there is a new state transition.  Events that are emitted after a state transition begins and before the next Enter() call are ignored, because at this point, next-state logic has already determined the next state.

> Is this class going to be used on any, including constrained, devices? I'm not sure how much attention I should pay to memory usage while reviewing :)

My intended first use is to rework some of the commissioner state machine.  It is likely that most commissioners are not constrained, but I understand that we can't exclude such implementations.  I think the biggest mitigation is that all of these allocations are short lived.  Queued events will always be immediately drained as the loop executes.  And there aren't any other dynamic allocation surprises in the PR.  This is the only one.

Reply:
> I think the biggest mitigation is that all of these allocations are short lived.

I suppose this may be no mitigation at all though depending on how std::deque is implemented.  I'll look at using something else.

Reply:
Also, thinking about this a bit more, the biggest problem is that this design has no answer for what happens if allocation for the queue fails.  Any allocation strategy would suffer this problem.

I'll come up with a solution.

Reply:
I've removed the queue entirely.  Instead, Dispatch now executes state logic immediately with preorder recursion.  This has cleaned things up quite a bit.  Dispatch can now be called from anywhere we like, including recursively, and will 'do the right thing'.

Acceptability of this solution relies upon the presumption that loops in the state machine must always be broken up by async operations or I/O.  That is not onerous at all though; that requirement already exists regardless of state machine implementation.